cmd/promtool/main.go

```diff
@@ -368,6 +375,52 @@ func checkTLSConfig(tlsConfig config_util.TLSConfig) error {
 	return nil
 }
 
+func checkSDFile(filename string) error {
+	fd, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	defer fd.Close()
+
+	content, err := ioutil.ReadAll(fd)
+	if err != nil {
+		return err
+	}
+
+	var targetGroups []*targetgroup.Group
+
+	switch ext := filepath.Ext(filename); strings.ToLower(ext) {
+	case ".json":
+		if err := json.Unmarshal(content, &targetGroups); err != nil {
+			return err
+		}
+	case ".yml", ".yaml":
+		if err := yaml.UnmarshalStrict(content, &targetGroups); err != nil {
+			return err
+		}
+	default:
+		return errors.Errorf("invalid file extension: %q", ext)
+	}
+
+	_, err = checkSDOutput(targetGroups)
+	return err
+}
+
+func checkSDOutput(targetGroups []*targetgroup.Group) ([]*targetgroup.Group, error) {
+	for i, tg := range targetGroups {
+		if tg == nil {
+			err := errors.Errorf("nil target group item found (index %v)", i)
+			return nil, err
+		}
+
+		if tg.Labels == nil {
+			tg.Labels = model.LabelSet{}
+		}
+		tg.Labels["__<filepath or url>"] = "<path or URL of your targets>"
```

Comment:
Why do we need this?

Reply:
I'm trying to replicate the SD code as much as possible so we give the most accurate representation of what will be available on the server.

https://github.com/prometheus/prometheus/blob/922f9babb1a6263b821cb879af4ed416c07d9469/discovery/file/file.go#L414
https://github.com/prometheus/prometheus/blob/922f9babb1a6263b821cb879af4ed416c07d9469/discovery/http/http.go#L179

Reply:
if we do this, we would probably use the SD interfaces, instead of doing  a per-sd thing.

Reply:
What do you mean by _per-sd thing_?

Reply:
I'm not sure this helps but I guess I'll throw in here that in the standalone validation command, which will actually use this output, we won't know if the input was from a file or from HTTP.

`curl localhost:1000 | promtool check sd`

`cat sd.yml | promtool check sd`

Either way Promtool receives a JSON or YAML encoded text blob.

Reply:
The idea of the validation command:


promtool check sd --config.file=prometheus --job=prometheus --wait-time=5s

--wait time is used is the SD is not a "refresh SD". It is the time we wait in the discovery channel to get the targets.

The output would be:

```
{targets: [ { job_name: prometheus, targets: [before_relabeling: {labels...}, after_relabeling: {labels....}]}
```

That enables to use oauth2, tls, etc, and works with any service discovery. it also enables testing relabeling.

WDYT?

Reply:
That's pretty cool, I really like it. 

My idea behind `curl localhost:1000 | promtool check sd` was that if you were developing a file or HTTP SD, you could quickly check your output without having to set up a config file or anything. I think that still might have a place, maybe if we don't get any flags we check if there's anything in stdin, and parse that. Maybe?

Reply:
there is more in the HTTP SD that the format. There is also the headers (content-type), the security, the return code ..

Reply:
True true, I guess we'll keep it at `promtool check sd --config.file=prometheus --job=prometheus --wait-time=5s`. For this PR though, I'll remove the `checkSDOutput` function now that we don't need it for anything else and move it into the `checkSDFile` function.

Reply:
We also don't need the output now.